html/mycakeapp/src/Controller/MypageController.php

```diff
@@ -42,4 +45,138 @@ public function delete()
+            $cardsInfoOwn[$key]['id'] = $encryptedCardId;
+        }
+        $title = '決済情報';
+        $this->set(compact('title', 'cardsInfoOwn'));
+    }
+    public function addpayment()
+    {
+        $this->viewBuilder()->setLayout('frame-title');
+        $entity = $this->Creditcards->newEntity();
+        $memberId = $this->Auth->user('id');
+        $today = date('Ymd');
+        if (!empty($this->request->query['id'])) { //update
+            // 暗号化したクレジットカードIDを戻す
+            $securityKey = Configure::read('key');
+            $securitySalt = Configure::read('salt');
+            $encryptedCardId = $this->request->query['id'];
+            $cardId = Security::decrypt($encryptedCardId, $securityKey, $securitySalt);
+            if ($cardId === false) { //不正なIDの時
+                return $this->redirect(['controller' => 'error']);
+            }
+            $cardHolderIsMember = $this->Creditcards->find()
+                ->where([
+                    'id' => $cardId,
+                    'member_id' => $memberId,
+                    'is_deleted' => 0,
+                    'deadline >=' => $today
+                ])->count();
+            if (!($cardHolderIsMember)) { //カードがユーザーのものじゃない時
+                return $this->redirect(['controller' => 'error']);
+            }
+
+            $entity = $this->Creditcards->get($cardId);
+            $entity = $this->Creditcards->patchEntity($entity, $this->request->getData());
+            //updateの日付のみ更新
+            if ($this->Creditcards->save($entity)) {
+                $_SESSION['addedpayment'] = 1;
+                return $this->redirect(['action' => 'addedpayment']);
+            }
+        } else { //insertのカード数判定
+            $numbeOfCardsOwned = $this->Creditcards->find()->where([
+                'member_id' => $this->Auth->user('id'),
+                'is_deleted' => 0,
+                'deadline >=' => $today
+            ])->count();
+            if ($numbeOfCardsOwned === 2) {
+                return $this->redirect(['controller' => 'error']);
+            }
+        }
+        if (!empty($this->request->is('post'))) { //insert
+            $entity = $this->Creditcards->patchEntity($entity, $this->request->getData());
+            $entity["member_id"] = $this->Auth->user('id');
+            $entity["is_deleted"] = 0;
+            $entity["created_at"] = date("Y/m/d H:i:s");
+            $entity["updated_at"] = date("Y/m/d H:i:s");
+            if ($this->Creditcards->save($entity)) {
+                $_SESSION['addedpayment'] = 1;
+                return $this->redirect(['action' => 'addedpayment']);
+            }
+        }
+        $title = "会員登録";
+        $this->set(compact('entity', 'title'));
+    }
+    public function addedpayment()
+    {
+        $this->viewBuilder()->setLayout('frame-title');
+        if (empty($_SESSION['addedpayment'])) {
+            return $this->redirect(['controller' => 'error']);
+        }
+        $_SESSION['addedpayment'] = null;
```

Comment:
この形だと$_SESSIONに['addedpayment']が値無しの状態残ってしまうので、
```suggestion
        $this->request->session()->delete('addedpayment');
```
の方が良いと思います！

Reply:
ありがとうございます！参考になります！！
こちらも修正いたします。